semantics/src/namespace/types.rs

```diff
@@ -9,37 +9,47 @@ pub trait FeSized {
     fn size(&self) -> usize;
 }
 
-/// The padding on an ABI type.
-pub enum AbiPadding {
-    /// The element is padded on the left with some number of bytes.
-    Left { size: usize },
-    /// The element is padded on the right with some number of bytes.
-    Right { size: usize },
-    /// There is no padding. For example, `u256` are already 32 bytes, so there
-    /// is no need to pad them.
-    None,
+#[derive(Clone, Debug, PartialEq, PartialOrd, Ord, Eq)]
+pub struct AbiUintSize {
+    pub data_size: usize,
+    pub padded_size: usize,
+}
+
+#[derive(Clone, Debug, PartialEq, PartialOrd, Ord, Eq)]
+pub enum AbiArraySize {
+    Static { size: usize },
+    Dynamic,
```

Comment:
create issue for dynamic types with max sizes. we should be able to attach a static max-size value here that is checked against the dynamic size during runtime. we should also allocate the max size when decoding